### PR TITLE
Document remove_almost_degenerate_faces()

### PR DESCRIPTION
**Pull request by sloriot**
## Summary of Changes

Document the function that used to be in the experimental namespace to remove almost degenerate faces. API has been update to put all parameters in a function named parameters

## Release Management

* Affected package(s): PMP
* Issue(s) solved (if any): NA
* Feature/Small Feature (if any): [here](https://cgal.geometryfactory.com/CGAL/Members/wiki/Features/Small_Features/RemoveCapsNeedles) pre-approved on 7/26/2022
* Link to compiled documentation (obligatory for small feature)   [link](https://cgal.github.io/6772/v0/Polygon_mesh_processing/group__PMP__repairing__grp.html#ga48008d2b66de8a68a7068f29db15dad6)
* License and copyright ownership: GF

### CODE REVIEW

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/repair_degeneracies.h [hunk 1]
```diff
@@ -536,15 +536,94 @@ struct Filter_wrapper_for_cap_needle_removal<TriangleMesh, VPM, Traits, Identity
...
+///   \cgalParamNEnd
+///   \cgalParamNBegin{vertex_is_constrained_map}
+///     \cgalParamDescription{a property map containing the constrained-or-not status of each vertex of `tmesh`.}
+///     \cgalParamType{a class model of `ReadWritePropertyMap` with `boost::graph_traits<PolygonMesh>::%vertex_descriptor`
+///                    as key type and `bool` as value type.}
+///     \cgalParamDefault{a default property map where no vertex is constrained.}
+///     \cgalParamExtra{A constrained vertex is guaranteed to be present in `tmesh` after the function call.}
+///   \cgalParamNEnd
+///   \cgalParamNBegin{filter}
+///     \cgalParamDescription{A function object providing `bool operator()('geom_traits::Point_3,geom_traits::Point_3,geom_traits::Point_3)`.}
```
Comment (afabri) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/repair_degeneracies.h:610]: The backtick before `geom_traits` is one too much.

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/repair_degeneracies.h [hunk 2]
```diff
@@ -536,15 +536,94 @@ struct Filter_wrapper_for_cap_needle_removal<TriangleMesh, VPM, Traits, Identity
...
+///     \cgalParamType{The function object is queried each time a new triangle is created by passing the three points and the flip or collapsed
```
Comment (afabri) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/repair_degeneracies.h:611]: "the flip or collapsed"  something is missing to make it a sentence.

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/repair_degeneracies.h [hunk 3]
```diff
@@ -536,15 +536,94 @@ struct Filter_wrapper_for_cap_needle_removal<TriangleMesh, VPM, Traits, Identity
...
+///                    responsible for the creation of that triangle is rejected if `false` is returned.}
+///     \cgalParamDefault{a functor always returning `true`.}
+///   \cgalParamNEnd
+/// \cgalNamedParamsEnd
+///
+/// \return `true` if no almost degenerate face could not be removed (due to topological constraints), and `false` otherwise.
```
Comment (afabri) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/repair_degeneracies.h:617]: Can we simplify the double negation?

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/repair_degeneracies.h [hunk 4]
```diff
@@ -536,15 +536,94 @@ struct Filter_wrapper_for_cap_needle_removal<TriangleMesh, VPM, Traits, Identity
 
 } // namespace internal
 
-namespace experimental {
-
-// @todo check what to use as priority queue with removable elements, set might not be optimal
+/// \ingroup PMP_repairing_grp
+///
+/// removes almost degenerate faces in a range of faces from a triangulated surface mesh.
+/// Almost degenerated triangle faces are classified as caps or needles: a triangle is said to be a needle<
```
Comment (afabri) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/repair_degeneracies.h:542]: Why not calling the function `remove_caps_and_needles()`.

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/repair_degeneracies.h [hunk 5]
```diff
@@ -536,15 +536,94 @@ struct Filter_wrapper_for_cap_needle_removal<TriangleMesh, VPM, Traits, Identity
...
+///                   and neighbor faces might also be impacted.
+/// @param tmesh the triangulated surface mesh to be modified
+/// @param np an optional sequence of \ref bgl_namedparameters "Named Parameters" among the ones listed below
+///
+/// \cgalNamedParamsBegin
+///   \cgalParamNBegin{cap_threshold}
+///     \cgalParamDescription{the cosine of a minimum angle such that if a face has an angle greater than this bound,
+///                           it is a cap. The threshold is in range `[-1 0]` and corresponds to an angle between `90` and `180` degrees.}
+///     \cgalParamType{`geom_traits::FT`}
+///     \cgalParamDefault{the cosinus corresponding to an angle of 160 degrees}
```
[the remark this answers is not in the document]
Comment (afabri) [reply Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/repair_degeneracies.h:565]: Ah, is 160 "almost 180"?

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/repair_degeneracies.h (hunk 3 above)
Comment (sloriot) [reply Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/repair_degeneracies.h:617]: That's on purpose to match both versions of the function, with a range of faces and all faces.

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/repair_degeneracies.h [hunk 6]
```diff
@@ -536,15 +536,94 @@ struct Filter_wrapper_for_cap_needle_removal<TriangleMesh, VPM, Traits, Identity
...
+/// Almost degenerated triangle faces are classified as caps or needles: a triangle is said to be a needle<
+/// if its longest edge is much longer than its shortest edge. A triangle is said to be a cap if one of
+/// its angles is close to `180` degrees. Needles are removed by collapsing their shortest edges, while caps are
+/// removed by flipping the edge opposite to the largest angle (with the exception of caps on the boundary that are
+/// simply removed from the mesh).
+///
+/// @pre `CGAL::is_triangle_mesh(tmesh)`
+///
+/// @tparam TriangleMesh a model of `FaceListGraph` and `MutableFaceGraph`
+/// @tparam a model of `ConstRange` with `boost::graph_traits<TriangleMesh>::%face_descriptor` as value type
```
Comment (afabri) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/repair_degeneracies.h:551]: missing variable name.

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/repair_degeneracies.h (hunk 4 above)
Comment (sloriot) [reply Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/repair_degeneracies.h:542]: Back when the function was created I think we decided to use the other name to be self explanatory. caps/needles is "just" terminology from some papers.

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/repair_degeneracies.h [hunk 7]
```diff
@@ -536,15 +536,94 @@ struct Filter_wrapper_for_cap_needle_removal<TriangleMesh, VPM, Traits, Identity
...
+///
+/// \cgalNamedParamsBegin
+///   \cgalParamNBegin{cap_threshold}
+///     \cgalParamDescription{the cosine of a minimum angle such that if a face has an angle greater than this bound,
+///                           it is a cap. The threshold is in range `[-1 0]` and corresponds to an angle between `90` and `180` degrees.}
+///     \cgalParamType{`geom_traits::FT`}
+///     \cgalParamDefault{the cosinus corresponding to an angle of 160 degrees}
+///   \cgalParamNEnd
+///   \cgalParamNBegin{needle_threshold}
+///     \cgalParamDescription{a bound on the ratio of the longest edge length and the shortest edge length such that a face having a ratio}
```
Comment (afabri) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/repair_degeneracies.h:568]: Incomplete sentence.

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/repair_degeneracies.h (hunk 4 above)
Comment (afabri) [reply Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/repair_degeneracies.h:542]: Well, that is how you define "almost".

**Review by afabri — APPROVED** [review]

**Comment by sloriot** [comment]
@MaelRL github seems to say that you have an on-going review. Do you have comments?

File: Polygon_mesh_processing/doc/Polygon_mesh_processing/Polygon_mesh_processing.txt [hunk 8]
```diff
@@ -835,6 +835,15 @@ more than once (although, with different vertices) before reaching the initial b
 `CGAL::Polygon_mesh_processing::merge_duplicated_vertices_in_boundary_cycle()`, which merge
 vertices at identical positions, can be used to repair this configuration.
 
+\subsection PMPRemoveCapsNeedles Removal of Almost Degenerate Triangle Faces
+Triangle faces of a mesh defined by almost collinear points are badly shaped elements that
```
Comment (MaelRL) [inline Polygon_mesh_processing/doc/Polygon_mesh_processing/Polygon_mesh_processing.txt:839]:
```suggestion
Triangle faces of a mesh made up of almost collinear points are badly shaped elements that
```

File: Polygon_mesh_processing/doc/Polygon_mesh_processing/Polygon_mesh_processing.txt [hunk 9]
```diff
@@ -835,6 +835,15 @@ more than once (although, with different vertices) before reaching the initial b
...
+might not be desirable to have in a mesh. The function
+`CGAL::Polygon_mesh_processing::remove_almost_degenerate_faces()` allows to remove such elements
```
Comment (MaelRL) [inline Polygon_mesh_processing/doc/Polygon_mesh_processing/Polygon_mesh_processing.txt:841]:
```suggestion
`CGAL::Polygon_mesh_processing::remove_almost_degenerate_faces()` enables removing such elements,
```

File: Polygon_mesh_processing/doc/Polygon_mesh_processing/Polygon_mesh_processing.txt [hunk 10]
```diff
@@ -835,6 +835,15 @@ more than once (although, with different vertices) before reaching the initial b
...
+using user parameters to quality what <i>almost</i> means (`cap_threshold` and `needle_threshold`).
```
Comment (MaelRL) [inline Polygon_mesh_processing/doc/Polygon_mesh_processing/Polygon_mesh_processing.txt:842]:
```suggestion
with user-defined parameters to qualify what <i>almost</i> means (`cap_threshold` and `needle_threshold`).
```

File: Polygon_mesh_processing/doc/Polygon_mesh_processing/Polygon_mesh_processing.txt [hunk 11]
```diff
@@ -835,6 +835,15 @@ more than once (although, with different vertices) before reaching the initial b
...
+As some badly shaped elements are inevitable (triangulation of a long cylinder
```
Comment (MaelRL) [inline Polygon_mesh_processing/doc/Polygon_mesh_processing/Polygon_mesh_processing.txt:843]:
```suggestion
As some badly shaped elements are inevitable (the triangulation of a long cylinder
```

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/repair_degeneracies.h [hunk 12]
```diff
@@ -536,15 +536,95 @@ struct Filter_wrapper_for_cap_needle_removal<TriangleMesh, VPM, Traits, Identity
...
+///
+/// @pre `CGAL::is_triangle_mesh(tmesh)`
+///
+/// @tparam TriangleMesh a model of `FaceListGraph` and `MutableFaceGraph`
+/// @tparam FaceRange a model of `ConstRange` with `boost::graph_traits<TriangleMesh>::%face_descriptor` as value type
+/// @tparam NamedParameters a sequence of \ref bgl_namedparameters "Named Parameters"
+///
+/// @param face_range the initial range of faces to be considered to look for badly shaped triangles.
+///                   Note that modifications of the `tmesh` is not limited to faces in `face_range`
+///                   and neighbor faces might also be impacted.
```
Comment (MaelRL) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/repair_degeneracies.h:556]:
```suggestion
///                   Note that modifications of `tmesh` are not limited to faces in `face_range`
///                   and neighbor faces might also be impacted.
```

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/repair_degeneracies.h [hunk 13]
```diff
@@ -536,15 +536,95 @@ struct Filter_wrapper_for_cap_needle_removal<TriangleMesh, VPM, Traits, Identity
...
+///     \cgalParamDefault{4}
+///   \cgalParamNEnd
+///   \cgalParamNBegin{collapse_length_threshold}
+///     \cgalParamDescription{if different from 0, an edge collapsed will be prevented if the edge is longer than the threshold given.}
+///     \cgalParamType{double}
+///     \cgalParamDefault{0}
+///   \cgalParamNEnd
+///   \cgalParamNBegin{flip_triangle_height_threshold}
+///     \cgalParamDescription{if different from 0, an edge flip will be prevented if the height of the triangle (with base being the edge to be flipped)
+///                           is longer than the threshold given.}
```
Comment (MaelRL) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/repair_degeneracies.h:580]:
```suggestion
///     \cgalParamDescription{if different from 0, an edge flip will be prevented if the height of the triangle (whose base is the edge to be flipped)
///                           is longer than the threshold given.}
```

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/repair_degeneracies.h [hunk 14]
```diff
@@ -536,15 +536,95 @@ struct Filter_wrapper_for_cap_needle_removal<TriangleMesh, VPM, Traits, Identity
...
+///   \cgalParamNEnd
+///   \cgalParamNBegin{vertex_point_map}
+///     \cgalParamDescription{a property map associating points to the vertices of `tmesh`.}
+///     \cgalParamType{a class model of `ReadablePropertyMap` with `boost::graph_traits<TriangleMesh>::%vertex_descriptor`
+///                    as key type and `%Point_3` as value type}
+///     \cgalParamDefault{`boost::get(CGAL::vertex_point, tmesh)`.}
+///   \cgalParamNEnd
+///   \cgalParamNBegin{geom_traits}
+///     \cgalParamDescription{an instance of a geometric traits class.}
+///     \cgalParamType{A model of `Kernel`.}
```
Comment (MaelRL) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/repair_degeneracies.h:592]: :hammer:

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/repair_degeneracies.h [hunk 15]
```diff
@@ -536,15 +536,95 @@ struct Filter_wrapper_for_cap_needle_removal<TriangleMesh, VPM, Traits, Identity
...
+///     \cgalParamDefault{a \cgal Kernel deduced from the point type, using `CGAL::Kernel_traits`.}
+///     \cgalParamExtra{The geometric traits class must be compatible with the vertex point type.}
+///   \cgalParamNEnd
+///   \cgalParamNBegin{edge_is_constrained_map}
+///     \cgalParamDescription{a property map containing the constrained-or-not status of each edge of `tmesh`.}
+///     \cgalParamType{a class model of `ReadWritePropertyMap` with `boost::graph_traits<PolygonMesh>::%edge_descriptor`
+///                    as key type and `bool` as value type.}
+///     \cgalParamDefault{a default property map where no edge is constrained.}
+///     \cgalParamExtra{A constrained edge can not be collapsed or flipped.}
```
Comment (MaelRL) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/repair_degeneracies.h:601]:
```suggestion
///     \cgalParamExtra{A constrained edge can not be collapsed nor flipped.}
```

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/repair_degeneracies.h [hunk 16]
```diff
@@ -536,15 +536,95 @@ struct Filter_wrapper_for_cap_needle_removal<TriangleMesh, VPM, Traits, Identity
...
+///                    as key type and `%Point_3` as value type}
+///     \cgalParamDefault{`boost::get(CGAL::vertex_point, tmesh)`.}
+///   \cgalParamNEnd
+///   \cgalParamNBegin{geom_traits}
+///     \cgalParamDescription{an instance of a geometric traits class.}
+///     \cgalParamType{A model of `Kernel`.}
+///     \cgalParamDefault{a \cgal Kernel deduced from the point type, using `CGAL::Kernel_traits`.}
+///     \cgalParamExtra{The geometric traits class must be compatible with the vertex point type.}
+///   \cgalParamNEnd
+///   \cgalParamNBegin{edge_is_constrained_map}
```
Comment (MaelRL) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/repair_degeneracies.h:596]: Not very important, but I don't think we need "writable" for ecm / vcm

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/repair_degeneracies.h [hunk 17]
```diff
@@ -536,15 +536,95 @@ struct Filter_wrapper_for_cap_needle_removal<TriangleMesh, VPM, Traits, Identity
...
+///     \cgalParamDescription{a property map containing the constrained-or-not status of each vertex of `tmesh`.}
+///     \cgalParamType{a class model of `ReadWritePropertyMap` with `boost::graph_traits<PolygonMesh>::%vertex_descriptor`
+///                    as key type and `bool` as value type.}
+///     \cgalParamDefault{a default property map where no vertex is constrained.}
+///     \cgalParamExtra{A constrained vertex is guaranteed to be present in `tmesh` after the function call.}
+///   \cgalParamNEnd
+///   \cgalParamNBegin{filter}
+///     \cgalParamDescription{A function object providing `bool operator()(geom_traits::Point_3,geom_traits::Point_3,geom_traits::Point_3)`.}
+///     \cgalParamType{The function object is queried each time a new triangle is created by passing the three points of the triangle.
+///                    The flip or collapse responsible for the creation of that triangle is rejected if `false` is returned.}
```
Comment (MaelRL) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/repair_degeneracies.h:613]:
```suggestion
///     \cgalParamType{The function object is queried each time a new triangle is about to be created by a flip or a collapse operation.
///                    If `false` is returned, the operation is cancelled.}
```

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/repair_degeneracies.h (hunk 3 above)
Comment (MaelRL) [reply Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/repair_degeneracies.h:618]: Symmetry is not worth the (small) tediousness to understand what this means, imo.

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/repair_degeneracies.h [hunk 18]
```diff
@@ -536,15 +536,95 @@ struct Filter_wrapper_for_cap_needle_removal<TriangleMesh, VPM, Traits, Identity
...
+///   \cgalParamNEnd
+///   \cgalParamNBegin{geom_traits}
+///     \cgalParamDescription{an instance of a geometric traits class.}
+///     \cgalParamType{A model of `Kernel`.}
+///     \cgalParamDefault{a \cgal Kernel deduced from the point type, using `CGAL::Kernel_traits`.}
+///     \cgalParamExtra{The geometric traits class must be compatible with the vertex point type.}
+///   \cgalParamNEnd
+///   \cgalParamNBegin{edge_is_constrained_map}
+///     \cgalParamDescription{a property map containing the constrained-or-not status of each edge of `tmesh`.}
+///     \cgalParamType{a class model of `ReadWritePropertyMap` with `boost::graph_traits<PolygonMesh>::%edge_descriptor`
```
Comment (sloriot) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/repair_degeneracies.h:598]:
```suggestion
///     \cgalParamType{a class model of `ReadablePropertyMap` with `boost::graph_traits<PolygonMesh>::%edge_descriptor`
```

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/repair_degeneracies.h [hunk 19]
```diff
@@ -536,15 +536,95 @@ struct Filter_wrapper_for_cap_needle_removal<TriangleMesh, VPM, Traits, Identity
...
+///                    as key type and `bool` as value type.}
+///     \cgalParamDefault{a default property map where no edge is constrained.}
+///     \cgalParamExtra{A constrained edge can not be collapsed or flipped.}
+///   \cgalParamNEnd
+///   \cgalParamNBegin{vertex_is_constrained_map}
+///     \cgalParamDescription{a property map containing the constrained-or-not status of each vertex of `tmesh`.}
+///     \cgalParamType{a class model of `ReadWritePropertyMap` with `boost::graph_traits<PolygonMesh>::%vertex_descriptor`
```
Comment (sloriot) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/repair_degeneracies.h:605]:
```suggestion
///     \cgalParamType{a class model of `ReadablePropertyMap` with `boost::graph_traits<PolygonMesh>::%vertex_descriptor`
```

**Comment by sloriot** [comment]
Successfully tested in CGAL-5.6-Ic-44

**Pull request merged**